### ISSUE TITLE
CIS AWS - Resource collection research/POC

### ISSUE THREAD

**Issue by ari-aviran** [elastic/cloudbeat#409]
To support CIS for AWS benchmark we need to collect data (resources) from different AWS services. We need to first do some research/POC on the best approach to collect said data. The kinds of data we'd like to collect (initial estimation by the product team, might change a bit):
- Enumerating and evaluating AWS resources for conformance to the rules in the [CIS AWS Foundations Benchmark v1.5.0](https://github.com/elastic/security-team/files/9603210/CIS_Amazon_Web_Services_Foundations_Benchmark_v1.5.0.pdf). The set of resources that will be enumerated and evaluated, in accordance with the benchmark, are limited to:
	- AWS Identity and Access Management (IAM)
	- IAM Access Analyzer
	- AWS Config
	- AWS CloudTrail
	- AWS CloudWatch
	- AWS Simple Notification Service (SNS)
	- AWS Simple Storage Service (S3)
	- Elastic Compute Cloud (EC2)
	- Relational Database Service (RDS)
	- AWS VPC
	
There are some 3rd-party tools that can assist in the data collection (e.g. `trivy`) that should be evaluated as part of this research.

**Out of scope**
- Building trust with AWS accounts (auth)
- Running cloudbeat in Elastic infra/hosted agent
- Shipping "raw" AWS resources to ES

**Comment by eyalkraft** [comment]
Thanks @ari-aviran!

Some additional notes:

### Motivation
As stated, We want to make sure we utilize existing tools that can save us precious time and effort regarding the collection of cloud resources and configuration from different cloud providers.
Since this domain is booming it's possible that a 3rd party tool (could require some effort to set up but overall) can save us a lot of maintenance in the future and would help us support new cloud providers and benchmarks faster in the future.

### Scope

The issue's emphasis is on resource collection, but another thing we should definitely include as part of the research is to find out whether we can also leverage the **policy evaluation** parts of these third party tools were assessing. For example Trivy [supports](https://blog.aquasec.com/now-you-can-scan-your-aws-services-for-security-issues-with-trivy) the AWS CIS benchmark. Could we convert Trivy's findings to our findings format and schema and then save ourselves the policy authoring work as well?

### 3rd part tools to evaluate

Some names to add to the list apart from [Trivy](https://github.com/aquasecurity/trivy) are [Cloudquery](https://github.com/cloudquery/cloudquery), [Cloudsploit](https://github.com/aquasecurity/cloudsploit) (Or is that just what's Trivy running under the hood?).

### DOD

To properly compare the tools and understand what we can benefit from them I suggest defining a table with some key parameters:
1. Ease/Cost of integration (Is it just calling a Go function vs Running severs/DBs/integrating other components/etc). This should be very detailed (What function should we call and how).
2. What's the benefit of using his tool?
    - Does It saves us from writing the fetchers? Which resources can we fetch using this tool?
    - Does the tool handle quotas, pagination, errors, rate limits, backoffs? If it does, how? Can we leverage it? Can we control it?
    - What are possible issues/upsides regarding the performance of this tool?
    - Does it save us from authoring policy? Does it produce resources or findings? Can we use the findings? Can we control the rules running?

Also worth adding the default option to the comparison - How hard will it be to implement the fetchers and the policy ourselves?

### Open questions

1. Are there AWS CIS Benchmark rules that talk about more than a single resource (and thus will require feeding OPA with more than a single resource at a time)?

**Comment by tinnytintin10** [comment]
Updated product epic: https://github.com/elastic/security-team/issues/5083

**Comment by tinnytintin10** [comment]
@ari-aviran @eyalkraft I see that the o11y team is doing a similar POC/feasibility analysis here https://github.com/elastic/obs-infraobs-team/issues/828 for sure worth looking into and sharing learnings!

**Comment by eyalkraft** [comment]
Additional thing to take into consideration with 3rd party tools:

- [ ] Are they multi-cloud? could the tool be beneficial for us also for the next benchmarks?

**Comment by DaveSys911** [comment]
@eyalkraft Another 3rd party worth taking into account is [cloud query by uptycs](https://github.com/uptycs/cloudquery).
Unlike its namesake, It's an osquery-based implementation - They support multi-cloud & other data sources like Kube-API.
They are an OSS company that offers a control plane saas built on top of OSquery - u can get the details here https://github.com/elastic/security-team/issues/1886.

I think it might be worth evaluating it as a data ingestion mechanism since elastic has already invested in Osquery itself.

**Comment by eyalkraft** [comment]
Interesting @DaveSys911!
Maybe we'll invest time in checking that out if @ari-aviran's research wouldn't bear fruit.
I do believe he has some interesting results he'll share soon.

**Comment by DaveSys911** [comment]
Understood.
Good idea to add a "licensing check" as part of the evaluation criteria. Some OSS tools might have more restrictive licensing that prevents just including them in our solution as is(like the elastic license).

**Comment by ari-aviran** [comment]
The scope of the research/POC here is utilizing tools or libraries that can collect AWS resources for us, to speed up development by not writing this logic ourselves. In addition, I've also looked at evaluation capabilities that some tools provide, in an effort to save more effort on our side and speed up the development even further.

I compared 3 OSS tools that can help us:
- [Trivy](https://github.com/aquasecurity/trivy) ([defsec](https://github.com/aquasecurity/defsec) library)
- [cloudquery](https://github.com/cloudquery/cloudquery) (platform)
- [cloudquery](https://github.com/Uptycs/cloudquery) (osquery extension)

Another tool, [cloudsploit](https://github.com/aquasecurity/cloudsploit), is written in JS and will be hard to use with cloudbeat, so I did not look into it all. 

Here is a table that summarizes the research:

Criteria \ Tool | Trivy (defsec library) | cloudquery (platform) | cloudquery (osquery extension)
-- | -- | -- | --
Fetching capabilities | Many AWS services - everything we need. | Many AWS services - everything we need. | Many AWS services - everything we need.
Evaluation capabilities | Partial implementation of CIS for AWS v1.2. Some rules are written in Rego and some in Go. We would probably need to write our own rules but we can take inspiration from their policy. | No evaluation capabilities. | No evaluation capabilities.
Ease of integration | Relatively easy. They expose a `Scanner` object as part of their Go API which we can initialize and commence fetching on demand. We can control which services we collect data from. | Not trivial. Cloudquery is a platform with plugins for “source” (where data comes from) and “destination” (where data is saved). There is an AWS source plugin but integrating with it from cloudbeat requires using high-level cloudquery API and semantics. | Not trivial. This is built as an osquery extension and there is no easy way to initialize data collection directly, it is all managed via osquery.
Integration concerns | <ul><li>No resource streaming - memory usage grows with the amount of cloud resources.</li><li>Auth with AWS handled internally in the library, we have no control over it.</li><li>Works per-region by default.</li></ul> | <ul><li>No resource streaming - memory usage grows with the amount of cloud resources.</li><li>Auth with AWS handled internally in the library, we have no control over it.</li><li>All-or-nothing services collection.</li></ul> | <ul><li>No resource streaming - memory usage grows with the amount of cloud resources.</li><li>Auth with AWS handled internally in the library, we have no control over it.</li><li>All-or-nothing services collection.</li></ul>
License | MIT. We can safely use this and modify if necessary | MPL-2. We can probably use it with minor limitations (if we change cloudquery source code we need to distribute them under the same MPL license). I am not a lawyer though. | Apache-2 / GPL (dual licensed). Can be used as Apache-2 so we can safely use it and modify if necessary
Popularity/maintenance | Relatively popular, regularly maintained | Somewhat popular, regularly maintained | Not popular, unmaintained

It seems that using Trivy (defsec) is the best option, and I went ahead and implemented a proof of concept for collecting S3 buckets and evaluating a single rule from the CIS for AWS benchmark. The POC went well and worked. Relevant links:
- [cloudbeat branch](https://github.com/elastic/cloudbeat/compare/main...ari-aviran:cloudbeat:cis_aws_poc) with the POC code.
- [csp-security-policies branch](https://github.com/elastic/csp-security-policies/compare/main...ari-aviran:csp-security-policies:cis_aws_poc) with the POC. code

**Comment by ari-aviran** [comment]
During implementation of the CIS AWS epic we have reverted our decision to use `defsec` and instead decided to implement resource collection on top of the AWS SDK. We have already found some disadvantages of using it, and when actually implementing on top of it we have found a couple more, which tipped the decision to not use it. In conclusion, here's what we didn't like with `defsec`:

- Holds all the resources together in memory.
- Not all resources are properly parsed.
- Fetches redundant data on some resources which we don't need for our benchmarks at the moment.
- Poor error handling (`defsec` isn't really designed as an SDK).
- Memory overhead (e.g. scanner per region).
- Data types that have redundant metadata at every layer and are not convenient to work with.
- Lack of control over collection code in case we want to make changes.
